File: sdk/go/run_helpers_test.go
package macrofold

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"
)

func finalResult(persistence, outcome string) string {
	return fmt.Sprintf(`{"run_id":%q,"final":true,"output_text":"Hello 🌍","execution_outcome":%q,"persistence_status":%q,"checkpoint_id":%q}`, testID, outcome, persistence, testID)
}
func helperState(status string) string {
	return fmt.Sprintf(`{"id":%q,"organization_id":%q,"session_id":%q,"workspace_id":%q,"harness":"codex","model":"fixture","status":%q,"failure_code":"fixture_failure","created_at":"2026-09-07T00:00:00Z"}`, testID, testID, testID, testID, status)
}

func TestTextFiltersReplayAndPreservesOrganization(t *testing.T) {
	streams, gets := 0, 0
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "GET" || r.Header.Get("X-Organization-Id") != testID {
			t.Error("read-only organization scope lost")
		}
		w.Header().Set("Content-Type", "application/json")
		switch {
		case strings.HasSuffix(r.URL.Path, "/stream"):
			streams++
			w.Header().Set("Content-Type", "text/event-stream")
			fmt.Fprintf(w, "data: %s\n\ndata: %s\n\n", event("1", "output.delta"), event("2", "tool.completed"))
			if streams > 1 {
				if r.Header.Get("Last-Event-ID") != "2" {
					t.Error("cursor lost")
				}
				fmt.Fprintf(w, "data: %s\n\ndata: %s\n\n", event("3", "reasoning.delta"), event("4", "run.succeeded"))
			}
		case strings.HasSuffix(r.URL.Path, "/result"):
			fmt.Fprint(w, finalResult("verified", "success"))
		default:
			gets++
			status := "running"
			if gets > 1 {
				status = "succeeded"
			}
			fmt.Fprint(w, helperState(status))
		}
	}))
	defer server.Close()
	client, _ := NewClient(WithBaseURL(server.URL), WithAPIKey("fixture"))
	var text []string
	err := client.Runs.StreamText(context.Background(), testID, "0", func(part string) error { text = append(text, part); return nil }, WithRequestOrganization(testID))
	if err != nil || strings.Join(text, "") != "hello 🌍" || len(text) != 1 || streams != 2 {
		t.Fatalf("text/replay: %v %v", text, err)
	}
}

func TestTextAfterTerminalAndWaitFailure(t *testing.T) {
	for _, status := range []string{"failed", "cancelled", "timed_out", "succeeded"} {
		t.Run(status, func(t *testing.T) {
			server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				if r.Method != "GET" {
					t.Error("unexpected mutation")
				}
				w.Header().Set("Content-Type", "application/json")
				switch {
				case strings.HasSuffix(r.URL.Path, "/stream"):
					return
				case strings.HasSuffix(r.URL.Path, "/events"):
					fmt.Fprint(w, `{"data":[],"next_cursor":null}`)
				case strings.HasSuffix(r.URL.Path, "/result"):
					persistence, outcome := "verified", status
					if status == "succeeded" {
						persistence, outcome = "failed", "success"
					}
					fmt.Fprint(w, finalResult(persistence, outcome))
				default:
					fmt.Fprint(w, helperState(status))
				}
			}))
			defer server.Close()
			client, _ := NewClient(WithBaseURL(server.URL), WithAPIKey("fixture"))
			err := client.Runs.StreamText(context.Background(), testID, "99", func(string) error { t.Fatal("unexpected text"); return nil })
			var failure *RunFailedError
			if !errors.As(err, &failure) || failure.RunID != testID || failure.Status != status || failure.FailureCode != "fixture_failure" {
				t.Fatalf("typed failure: %v", err)
			}
		})
	}
}

func TestWaitPersistenceAndDetachment(t *testing.T) {
	gets, results := 0, 0
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "GET" {
			t.Error("unexpected cancellation")
		}
		w.Header().Set("Content-Type", "application/json")
		switch {
		case strings.HasSuffix(r.URL.Path, "/stream"):
			w.Header().Set("Content-Type", "text/event-stream")
			fmt.Fprintf(w, "data: %s\n\n", event("1", "output.delta"))
		case strings.HasSuffix(r.URL.Path, "/result"):
			results++
			persistence := "pending"
			if results > 1 {
				persistence = "verified"
			}
			fmt.Fprint(w, finalResult(persistence, "success"))
		default:
			gets++
			status := "persisting"
			if gets > 1 {
				status = "succeeded"
			}
			fmt.Fprint(w, helperState(status))
		}
	}))
	defer server.Close()
	client, _ := NewClient(WithBaseURL(server.URL), WithAPIKey("fixture"))
	detached := errors.New("detached")
	if err := client.Runs.StreamText(context.Background(), testID, "0", func(string) error { return detached }); err != detached {
		t.Fatalf("detach: %v", err)
	}
	if gets != 0 || results != 0 {
		t.Fatal("detach performed completion reads")
	}
	value, err := client.Runs.Wait(context.Background(), testID)
	if err != nil || value.GetCheckpointId() != testID || gets != 3 || results != 2 {
		t.Fatalf("persistence: %v %v", value, err)
	}
}

func TestWaitTimeoutAbortsInflightWithoutCancelling(t *testing.T) {
	calls := 0
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		calls++
		if r.Method != "GET" {
			t.Error("unexpected cancellation")
		}
		<-r.Context().Done()
	}))
	defer server.Close()
	client, _ := NewClient(WithBaseURL(server.URL), WithAPIKey("fixture"))
	ctx, cancel := context.WithTimeout(context.Background(), 50*time.Millisecond)
	defer cancel()
	_, err := client.Runs.Wait(ctx, testID)
	var failure *WaitTimeoutError
	if !errors.As(err, &failure) || failure.RunID != testID || !errors.Is(err, context.DeadlineExceeded) {
		t.Fatalf("timeout: %v", err)
	}
	if calls != 1 {
		t.Fatalf("unexpected requests: %d", calls)
	}
}

File: sdk/java/src/test/java/dev/macrofold/RunHelpersTest.java
package dev.macrofold;

import static org.junit.jupiter.api.Assertions.*;
import com.sun.net.httpserver.HttpServer;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import org.junit.jupiter.api.Test;

class RunHelpersTest {
  static final UUID ID = UUID.fromString(ClientTest.ID);
  static String state(String status) {
    return "{\"id\":\"" + ID + "\",\"status\":\"" + status + "\",\"failure_code\":\"fixture_failure\"}";
  }
  static String result(String persistence, String outcome) {
    return "{\"run_id\":\"" + ID + "\",\"final\":true,\"output_text\":\"Hello 🌍\",\"execution_outcome\":\"" + outcome + "\",\"persistence_status\":\"" + persistence + "\",\"checkpoint_id\":\"" + ID + "\"}";
  }
  static final class Fixture implements AutoCloseable {
    final HttpServer server;
    final Client client;
    final List<String> paths = new ArrayList<>(), organizations = new ArrayList<>(), cursors = new ArrayList<>();
    Fixture(String... responses) throws Exception {
      server=HttpServer.create(new InetSocketAddress("127.0.0.1",0),0);
      server.createContext("/", exchange -> {
        paths.add(exchange.getRequestMethod()+" "+exchange.getRequestURI());
        organizations.add(exchange.getRequestHeaders().getFirst("X-Organization-Id"));
        cursors.add(exchange.getRequestHeaders().getFirst("Last-Event-ID"));
        boolean expected=paths.size()<=responses.length;
        String response=expected?responses[paths.size()-1]:"{}";
        byte[] bytes=response.getBytes(StandardCharsets.UTF_8);
        exchange.getResponseHeaders().set("Content-Type",response.startsWith("data:")?"text/event-stream":"application/json");
        exchange.sendResponseHeaders(expected?200:500,bytes.length);
        try(var out=exchange.getResponseBody()){out.write(bytes);}
      });
      server.start();
      client=new Macrofold("http://127.0.0.1:"+server.getAddress().getPort(),"fixture");
    }
    public void close(){server.stop(0);}
  }
  @Test void textFiltersReconnectsAndKeepsOrganization() throws Exception {
    try(var fixture=new Fixture(
        ClientTest.event("1","output.delta")+ClientTest.event("2","tool.completed"),state("running"),
        ClientTest.event("1","output.delta")+ClientTest.event("2","tool.completed")+ClientTest.event("3","reasoning.delta")+ClientTest.event("4","run.succeeded"),
        state("succeeded"),result("verified","success"))) {
      List<String> text=new ArrayList<>();
      fixture.client.runs().withOptions(new RequestOptions(null,ID)).streamText(ID, part->{text.add(part);return true;});
      assertEquals(List.of("hello 🌍"),text);
      assertEquals("2",fixture.cursors.get(2));
      assertEquals(Collections.nCopies(5,ID.toString()),fixture.organizations);
      assertTrue(fixture.paths.stream().allMatch(path->path.startsWith("GET ")));
    }
  }
  @Test void terminalReplayRaisesTypedFailures() throws Exception {
    for(String status:List.of("failed","cancelled","timed_out","succeeded")) {
      try(var fixture=new Fixture("",state(status),"{\"data\":[],\"next_cursor\":null}",state(status),
          result(status.equals("succeeded")?"failed":"verified",status.equals("succeeded")?"success":status))) {
        RunFailedException error=assertThrows(RunFailedException.class,()->fixture.client.runs().streamText(ID,"99",text->fail("Unexpected text")));
        assertEquals(ID,error.getRunId());
        assertEquals(status,error.getStatus());
        assertEquals("fixture_failure",error.getFailureCode());
        assertTrue(error.getMessage().contains(ID.toString()));
      }
    }
  }
  @Test void waitIncludesPersistenceAndDetachmentSkipsCompletion() throws Exception {
    try(var fixture=new Fixture(ClientTest.event("1","output.delta"),state("persisting"),state("succeeded"),result("pending","success"),state("succeeded"),result("verified","success"))) {
      fixture.client.runs().streamText(ID,text->false);
      assertEquals(1,fixture.paths.size());
      var result=fixture.client.runs().wait(ID);
      assertEquals("Hello 🌍",result.getOutputText());
      assertEquals(ID,result.getCheckpointId());
      assertEquals(6,fixture.paths.size());
      assertTrue(fixture.paths.subList(1,6).stream().noneMatch(path->path.contains("/stream")));
    }
  }
  @Test void timeoutClosesSlowBodyAndNeverCancelsExecution() throws Exception {
    HttpServer server=HttpServer.create(new InetSocketAddress("127.0.0.1",0),0);
    CountDownLatch release=new CountDownLatch(1), reached=new CountDownLatch(1);
    List<String> methods=new ArrayList<>();
    server.createContext("/",exchange->{
      methods.add(exchange.getRequestMethod());
      exchange.sendResponseHeaders(200,1000);
      exchange.getResponseBody().write('{');exchange.getResponseBody().flush();reached.countDown();
      try {release.await(5,TimeUnit.SECONDS);} catch(InterruptedException error){Thread.currentThread().interrupt();}
      finally{exchange.close();}
    });
    server.start();
    try {
      var client=new Macrofold("http://127.0.0.1:"+server.getAddress().getPort(),"fixture");
      var error=assertThrows(WaitTimeoutException.class,()->client.runs().wait(ID,Duration.ofMillis(500)));
      assertEquals(ID,error.getRunId());
      assertTrue(reached.await(1,TimeUnit.SECONDS));
      assertEquals(List.of("GET"),methods);
      assertThrows(WaitTimeoutException.class,()->client.runs().wait(ID,Duration.ZERO));
      assertEquals(1,methods.size());
      assertThrows(IllegalArgumentException.class,()->client.runs().wait(ID,Duration.ofSeconds(-1)));
    } finally {release.countDown();server.stop(0);}
  }
}
